Blankbot cmake modernization

jlknauer opened the pull request:
Depends on https://github.com/cpp-sc2/cpp-sc2/pull/118

Not for immediate merging due to dependency, and the `GIT_REPOSITORY` and `GIT_TAG` for `cpp-sc2` are currently pointing to my fork commit for the dependent PR. Once the `cpp-sc2` PR is approved and merged, I will rebase the correct `FetchContent` configuration.

Otherwise this PR is for review of changes made to bring `BlankBot` in line with the incoming `cpp-sc2` build configuration changes.

jrtknauer committed:
build: cmake dependency approach refactor

Extending the work completed in cpp-sc2 for switching from git
submodules to CMake FetchContent for managing dependencies.

jrtknauer committed:
docs: updated build documentation

jrtknauer committed:
build: wsl2 cross compilation support with cpp-sc2

jrtknauer committed:
docs: wsl2 cross compilation documentation

ghost commented:
Waiting on https://github.com/cpp-sc2/cpp-sc2/pull/121 to be merged to set the `cpp-sc2` `FetchContent` commit. Once rebased the PR will be ready for merge.

alkurbatov commented on cmake/cpp_sc2.cmake:
```diff
@@ -0,0 +1,22 @@
+message(STATUS "FetchContent: cpp-sc2")
```
Please use the same folders structure as in cpp-sc2, i.e. this one should be thirdparty/... as cpp-sc2 is external dependency for this bot.

alkurbatov commented on src/CMakeLists.txt:
```diff
@@ -2,41 +2,28 @@
-    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -Wextra -pedantic")
-endif ()
+    target_compile_options(BlankBot PRIVATE -Wall -Wextra -pedantic)
+endif()
```
Should be `endif ()`.

alkurbatov commented on src/CMakeLists.txt:
```diff
@@ -2,41 +2,28 @@
-
-# Linux specific.
-if (UNIX AND NOT APPLE)
+elseif (UNIT AND NOT APPLE)
```
Typo: `UNIT` -> `UNIX`.

Also:
* If we use `elseif` there is no need to doublecheck `NOT APPLE`.
* Please keep the original comment regarding Linux, it is kept mostly for people less experienced with cmake.

alkurbatov commented on CMakeLists.txt:
```diff
@@ -4,27 +4,43 @@
-# Dont build tests in the sc2api submodule.
-set(BUILD_API_TESTS OFF CACHE INTERNAL "" FORCE)
+if (MSVC)
+    # Setup MSVC parallelized builds
```
Style: Please, finish the comments with dots for consistency (or drop dots everywhere).
Here and in the other places.

alkurbatov commented on README.md:
````diff
@@ -104,19 +106,37 @@ Target SC2 version: 5.0.5
 always behind the Windows version. It is possible by specifying the game version manually, e.g.:
 ```bash
-$ cmake -DSC2_VERSION=4.10.0 ../
+$ cmake -DSC2_VERSION=4.10.0
````
In this case, examples should use `-B build` as we are outside of the build folder.
Here and below.

alkurbatov commented on README.md:
```diff
@@ -36,63 +36,65 @@ Feel free to post questions on the unofficial StarCraft II AI Discord server. [I
-# Use 'cmake -DCMAKE_BUILD_TYPE=Debug ../' if debuginfo is needed
-$ cmake ../
+# Generate CMake build tree
+# Options:
```
I believe this list of options is unnecessary because we already have detailed description below.
I would prefer to keep it in one place to reduce the maintenance cost.

alkurbatov reviewed (changes requested):
Looks good but needs minor corrections.

alkurbatov commented:
Btw, it seems that the PR has passed on ubuntu-18.04. Did they decided to extend support of the distro?

ghost commented:
Yes as per [the original announcement](https://github.com/actions/runner-images/issues/6002), actions will fail during brownouts (like on the original deprecation start date) and the image will be fully deprecated in December.

jrtknauer committed:
build: final cmake cleanup

Build structure normalized to match cpp-sc2, along with some minor style
changes and fixes.

alkurbatov commented on README.md:
````diff
@@ -36,87 +36,99 @@ Feel free to post questions on the unofficial StarCraft II AI Discord server. [I
 To compile a bot capable to play on [the AIArena ladder](https://aiarena.net), configure the project in the following way:
 ```bash
-$ cmake -DBUILD_FOR_LADDER=ON -DSC2_VERSION=4.10.0 ../
+$ cmake -DBUILD_FOR_LADDER=ON -DSC2_VERSION=4.10.0
````
Missing `-B build`.

alkurbatov commented:
Looks very good, only one fix  left (in README).

alkurbatov commented on README.md:
````diff
@@ -36,87 +36,99 @@ Feel free to post questions on the unofficial StarCraft II AI Discord server. [I
 To compile a bot capable to play on [the AIArena ladder](https://aiarena.net), configure the project in the following way:
 ```bash
-$ cmake -DBUILD_FOR_LADDER=ON -DSC2_VERSION=4.10.0 ../
+$ cmake -B -DBUILD_FOR_LADDER=ON -DSC2_VERSION=4.10.0
````
Looks like the `build` part is still missing.

ghost commented:
Maybe hold off on merge to rebase `cpp-sc2` commit hash to https://github.com/cpp-sc2/cpp-sc2/pull/122 once merged, as WSL2 options will fail here as well on CMake versions older than 3.24.

alkurbatov commented:
Thank you! Very good work.
